cmd/ttn-lw-cli/commands/login.go

```diff
@@ -31,40 +35,76 @@ var (
 		Use:   "login",
 		Short: "Login",
 		RunE: func(cmd *cobra.Command, args []string) error {
-			lis, err := net.Listen("tcp", ":11885")
+			ctx, done := context.WithCancel(ctx)
+			defer done()
+
+			callback, err := cmd.Flags().GetBool("callback")
 			if err != nil {
 				return err
 			}
-			var (
-				once    sync.Once
-				tokenCh = make(chan *oauth2.Token)
-			)
-			go http.Serve(lis, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-				if r.Method != http.MethodGet || r.URL.Path != "/oauth/callback" {
-					http.NotFound(w, r)
-					return
-				}
-				token, err := oauth2Config.Exchange(ctx, r.URL.Query().Get("code"))
-				if err != nil {
-					msg := "Could not exchange OAuth access token"
-					logger.WithError(err).Error(msg)
-					w.WriteHeader(http.StatusUnauthorized)
+
+			var token *oauth2.Token
+
+			if callback {
+				oauth2Config.RedirectURL = "http://localhost:11885/oauth/callback"
+
+				http.HandleFunc("/oauth/callback", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					if r.Method != http.MethodGet {
+						http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+						return
+					}
+					token, err = oauth2Config.Exchange(ctx, r.URL.Query().Get("code"))
+					if err != nil {
+						msg := "Could not exchange OAuth access token"
+						logger.WithError(err).Error(msg)
+						w.WriteHeader(http.StatusUnauthorized)
+						w.Write([]byte(msg))
+						return
+					}
+					msg := "Got OAuth access token"
+					logger.Info(msg)
 					w.Write([]byte(msg))
-					return
+					done()
+				}))
+
+				lis, err := net.Listen("tcp", ":11885")
+				defer lis.Close()
+				if err != nil {
+					return err
 				}
-				msg := "Got OAuth access token"
-				logger.Info(msg)
-				w.Write([]byte(msg))
-				once.Do(func() {
-					tokenCh <- token
-				})
-			}))
+				go http.Serve(lis, nil)
+			} else {
+				oauth2Config.RedirectURL = "/oauth/code"
+			}
 
 			logger.Infof("Please go to %s", oauth2Config.AuthCodeURL(""))
-			logger.Info("Waiting for your authorization...")
 
-			token := <-tokenCh
-			lis.Close()
+			if callback {
+				logger.Info("Waiting for your authorization...")
+				<-ctx.Done()
+			} else {
+				logger.Info("Then paste the authorization code and press enter")
+				var code string
+				for {
```

Comment:
Why the `for` loop here?

Reply:
I wanted to add some kind of validation here, but then forgot to finish it. I'll push that in a couple of minutes.

Reply:
This still seems overly complicated for it's purpose.
Why not just read 1 line and fail on errors?(the user can then just run `ttn-lw-cli login --callback=false` again)
What problem do we solve here by looping and continuing on errors? That's certainly not a behavior I would expect to see as a user.
Besides, the use case is that user copy-pastes the code - why would the input contain multiple lines? If it does contain multiple lines - it's certainly incorrect - should we really be trying every single line of input we already know is wrong hoping to get a "match"?